src/metabase/api/database.clj

```diff
@@ -404,54 +404,106 @@
 
-(defn- autocomplete-tables [db-id search-string limit]
+(defn- match-search-string [match-type search-string]
+  (str/lower-case
+   (case match-type
+     :prefix (str search-string "%")
+     :substring (str "%" search-string "%"))))
+
+(defn- autocomplete-tables [db-id search-string limit match-type]
   (db/select [Table :id :db_id :schema :name]
     {:where    [:and [:= :db_id db-id]
                      [:= :active true]
-                     [:like :%lower.name (str/lower-case search-string)]
+                     [:like :%lower.name (match-search-string match-type (str/lower-case search-string))]
                      [:= :visibility_type nil]]
      :order-by [[:%lower.name :asc]]
      :limit    limit}))
 
-(defn- autocomplete-fields [db-id search-string limit]
+(defn- autocomplete-fields [db-id search-string limit match-type]
   (db/select [Field :name :base_type :semantic_type :id :table_id [:table.name :table_name]]
     :metabase_field.active          true
-    :%lower.metabase_field.name     [:like (str/lower-case search-string)]
+    :%lower.metabase_field.name     [:like (match-search-string match-type (str/lower-case search-string))]
     :metabase_field.visibility_type [:not-in ["sensitive" "retired"]]
     :table.db_id                    db-id
     {:order-by  [[:%lower.metabase_field.name :asc]
                  [:%lower.table.name :asc]]
      :left-join [[:metabase_table :table] [:= :table.id :metabase_field.table_id]]
      :limit     limit}))
 
-(defn- autocomplete-results [tables fields limit]
-  (let [tbl-count   (count tables)
-        fld-count   (count fields)
-        take-tables (min tbl-count (- limit (/ fld-count 2)))
-        take-fields (- limit take-tables)]
-    (concat (for [{table-name :name} (take take-tables tables)]
-              [table-name "Table"])
-            (for [{:keys [table_name base_type semantic_type name]} (take take-fields fields)]
-              [name (str table_name
-                         " "
-                         base_type
-                         (when semantic_type
-                           (str " " semantic_type)))]))))
-
-(defn- autocomplete-suggestions
-  "match-string is a string that will be used with ilike. The it will be lowercased by autocomplete-{tables,fields}. "
-  [db-id match-string]
-  (let [limit  50
-        tables (filter mi/can-read? (autocomplete-tables db-id match-string limit))
-        fields (readable-fields-only (autocomplete-fields db-id match-string limit))]
-    (autocomplete-results tables fields limit)))
+(defn- autocomplete-card-columns
+  "Returns a list of card column metadata given the `search-string` and a set of card ids.
+   The search-string is matched against column names in the result_metadata of the cards.
+  Each result contains all the metadata for the column, and the name of the card."
+  [db-id search-string limit match-type card-ids]
```

Comment:
```suggestion
  [db-id search-string card-ids limit match-type]
```
nit: I think `card-ids` can be considered as more important than `limit` and `match-type` and it's essential for search, so I'd move it up a bit.